tty/uart16550/uart16550.c

```diff
@@ -129,46 +134,99 @@ static void set_cflag(void *_uart, tcflag_t *cflag)
 static void signal_txready(void *arg)
 {
 	uart_t *uart = (uart_t *)arg;
-
-	uarthw_write(uart->hwctx, REG_IMR, IMR_THRE | IMR_DR);
 	condSignal(uart->intcond);
 }
 
 #ifdef __TARGET_RISCV64
-__attribute__((section(".interrupt"), aligned(0x1000)))
-#endif
-static int
-uart_interrupt(unsigned int n, void *arg)
+__attribute__((section(".interrupt"), aligned(0x1000))) static int uart_interrupt(unsigned int n, void *arg)
 {
-	return ((uart_t *)arg)->intcond;
+	/* RISC-V platform is very special in how it handles memory during interrupts.
+	 * Due to this the uart_interrupt function cannot call uarthw_* functions.
+	 * Fortunately the UART IRQ on this platform is edge-triggered so we can just
+	 * exit interrupt, signal uart->intcond and main thread will handle the rest.
+	 */
+	uart_t *uart = (uart_t *)arg;
+	return uart->intcond;
 }
+#else
+static int uart_interrupt(unsigned int n, void *arg)
+{
+	uart_t *uart = (uart_t *)arg;
+
+	/* Caution: implementation-dependent behavior!
+	 * On some UARTs masking interrupts after an interrupt has been asserted
+	 * does not de-assert the IRQ line. In this case it is necessary to handle
+	 * the interrupt's cause fully within the ISR (e.g. reading the whole FIFO).
+	 * On other UARTs masking interrupts de-asserts IRQ and changes the value of IIR.
+	 * To handle this case we read IIR before masking interrupts. Note that in this
+	 * case the FIFO will not be fully read within the ISR.
+	 */
+	uint8_t iir = uarthw_read(uart->hwctx, REG_IIR);
+	uarthw_write(uart->hwctx, REG_IMR, 0);
+	unsigned int i = uart->buf_i;
+	do {
+		uint8_t intr_type = (iir >> 1) & 0x7;
+		if ((intr_type == IIR_CODE_DR) || (intr_type == IIR_CODE_RTO)) {
+			uint8_t c = uarthw_read(uart->hwctx, REG_RBR);
+			if (i < SW_BUF_SIZE) {
+				uart->buf[i] = c;
+				i++;
+			}
+		}
+		else if (intr_type == IIR_CODE_LS) {
+			uarthw_read(uart->hwctx, REG_LSR);
+		}
+		else if (intr_type == IIR_CODE_MS) {
+			uarthw_read(uart->hwctx, REG_MSR);
+		}
+
+		iir = uarthw_read(uart->hwctx, REG_IIR);
+	} while ((iir & IIR_IRQPEND) == 0);
+
+	uart->buf_i = i;
+	return uart->intcond;
+}
+#endif
 
 
 static void uart_intthr(void *arg)
 {
 	uart_t *uart = (uart_t *)arg;
-	uint8_t iir;
+	uint8_t target_imr = IMR_DR;
 
 	mutexLock(uart->mutex);
 	for (;;) {
-		while ((iir = uarthw_read(uart->hwctx, REG_IIR)) & IIR_IRQPEND) {
-			condWait(uart->intcond, uart->mutex, 0);
+		uarthw_write(uart->hwctx, REG_IMR, target_imr);
+		condWait(uart->intcond, uart->mutex, 0);
+		/* For the following part the interrupt needs to be masked */
+		uarthw_write(uart->hwctx, REG_IMR, 0);
+
+		/* Empty received buffer */
+		unsigned int buf_i = uart->buf_i;
+		for (unsigned int i = 0; i < buf_i; i++) {
```

Comment:
I assume that the compiler will optimize-out this loop on RISC-V, as `uart->buf_i` is always 0, so it is not worth to `#ifdef` this loop?